### PR TITLE
Startup leader election

### LINKED ISSUE #354: Changing Record page experiment records upon refresh
**Bug Description**
Root Cause: 5 Concurrent Startup Syncs Racing                                                                                  
                                                                                                                                 
  The CloudWatch logs reveal a clear race condition. When the ECS task b2bc38b89bda4d8d9a72e3d85069fe7c started at 17:03, all 5  
  Uvicorn workers launched run_startup_sync() simultaneously:                                                                    
                                                                                                                                 
  pid:66 - Starting one-time startup sync (20/20 experiments missing locally)                                                    
  pid:68 - Starting one-time startup sync (20/20 experiments missing locally)                                                    
  pid:69 - Starting one-time startup sync (20/20 experiments missing locally)                                                    
  pid:67 - Starting one-time startup sync (20/20 experiments missing locally)                                                    
  pid:70 - Starting one-time startup sync (20/20 experiments missing locally)                                                    
                                                                                                                                 
  All 5 workers see the same 20 experiments missing and start downloading the same files from S3 to the same local disk paths
  concurrently:
  pid:66: downloading 113/daa73197 -> thumbnails_only
  pid:68: downloading 113/daa73197 -> thumbnails_only  ← SAME experiment!
  pid:69: downloading 113/daa73197 -> thumbnails_only  ← SAME experiment!

### Why Records Flicker

  The GET /experiments/{workspace_id} endpoint (experiment.py:41) reads experiment data from the local filesystem via glob():

  1. Files appear mid-write: glob() finds an experiment.yaml that's currently being written by a sync process. The read gets
  partial/empty content ({}), validate_experiment_config() rejects it → experiment disappears from the response.
  2. Inconsistent disk state between refreshes: Each time you refresh, the filesystem is in a different state as the 5 workers
  download overlapping files. Some experiments may have their YAML fully written, others not yet.
  3. The evidence - recalculate_workspace_data_capacity_with_s3 ran twice for workspace 85 within 23 seconds:
    - 17:04:15 (pid:68): 6 experiment records
    - 17:04:38 (pid:69): 7 experiment records ← one more file appeared on disk
  4. Corrupted experiment.yaml files: The logs show many warnings:
  Skipping YAML update for experiment 111/tutorial4: Invalid or corrupted
  experiment.yaml file ([{}])
  4. These {} files fail validation and cause records to disappear.

  Additional Sync Activity Piling On

  On top of the 5 concurrent startup syncs, there's ALSO:
  - User login sync (sync_user_experiments) triggered at 17:22:33 for user 2 (premium)
  - Data capacity recalculation triggered by your client 9463168e798ef993 when browsing workspaces

  All of these are writing to the same local filesystem concurrently with no coordination.

### The Fix Needed

  The startup sync in studio/__main_unit__.py:112 needs a leader election so only one worker runs it. Currently every worker's
  lifespan triggers it independently.

**Priority**
Classify the severity of this bug as [Low, Medium, High, Critical].
- Low: Minimal impact on functionality, can be fixed in a future release.

### PR DESCRIPTION
**Pull request by milesAraya**
##### Backend
- `studio/__main_unit__.py` -- Add leader election guard to `_startup_sync()` so only 1 of N workers runs startup sync
- `studio/app/common/core/storage/atomic_claim_file.py` -- New: generic atomic file claim primitive using `O_CREAT|O_EXCL` with PID liveness, hostname comparison, and configurable expiry
- `studio/app/common/core/storage/startup_leader.py` -- New: thin wrapper around `AtomicClaimFile` for startup leader election
- `studio/app/common/routers/users_me.py` -- Replace `db.add()`/`db.commit()` with direct SQL `db.execute(update(...))` to avoid `StaleDataError` across concurrent workers
- `studio/app/common/core/middleware/user_activity_middleware.py` -- Same concurrent-write safety fix for `clear_free_user_logged_out_at` and `_update_free_user_activity_sync`; add `IntegrityError` handling for insert race conditions

##### Tests
- `studio/tests/app/common/core/storage/test_atomic_claim_file.py` -- New: 25 tests covering atomic creation, staleness detection (expired, dead PID, hostname mismatch, corrupt JSON), release, and stale-cleanup retry
- `studio/tests/app/common/core/storage/test_startup_leader.py` -- New: 10 tests covering leader election, release lifecycle, stale leader cleanup, and time-based expiry
- `studio/tests/app/test_main_unit_startup.py` -- New: 3 integration tests verifying leader runs sync, non-leader skips, and release happens on error
- `studio/tests/app/common/routers/test_users_me_logout.py` -- Update mocks to match new `db.execute(update(...))` return pattern (`rowcount` instead of `first()`)

#### Design Decisions
- **File-based leader election instead of Redis or DB lock.** The system runs on a single EBS volume shared by all uvicorn workers in the same container. `O_CREAT|O_EXCL` is atomic on POSIX filesystems and requires no additional infrastructure. If the claim file mechanism fails for any reason, all workers fall through to the original behavior (safe degradation).
- **15-minute expiry on leader claim.** Startup sync typically completes in 1-2 minutes. A 15-minute window provides generous headroom while still allowing recovery if the leader process crashes without releasing.
- **PID liveness + hostname checks for staleness.** A leftover claim file from a previous container (different hostname) or a dead process (PID no longer alive) is automatically cleaned up on the next election attempt. This handles ECS task recycling without manual intervention.
- **Direct SQL updates in users_me.py and middleware.** The previous pattern (`session.query().first()` then `obj.field = value; session.add()`) is vulnerable to `StaleDataError` when multiple workers read and write the same row concurrently. `db.execute(update(...).where(...).values(...))` is a single atomic SQL statement that avoids this.
- **IntegrityError catch in `_update_free_user_activity_sync`.** A narrow race exists where two workers both see `rowcount == 0` on UPDATE and both attempt INSERT. Catching `IntegrityError` and treating it as success handles this benign race without a distributed lock.

### References
- `DATA_DOWNLOAD_REPORT.md` -- Issue B: "5 uvicorn workers all run startup sync simultaneously"
- `DATA_DOWNLOAD_PLAN_2.md` -- Staged PR plan, PR 1 specification

### Manual Testcases
- Deploy to staging with 5 uvicorn workers
- Check logs: only 1 PID should show `coordinator.leader_elected`; other 4 should show `coordinator.leader_deferred`
- Verify all published experiments appear on the Records page after startup completes
- Kill the leader worker mid-sync, then restart -- a new worker should win election (stale PID detection)
- Cycle the ECS task -- new container should elect a new leader (stale hostname detection)
- `pytest studio/tests/app/common/core/storage/test_atomic_claim_file.py studio/tests/app/common/core/storage/test_startup_leader.py studio/tests/app/test_main_unit_startup.py studio/tests/app/common/routers/test_users_me_logout.py -v` -- all 43 tests pass

### Unit, Integration, Contract Test Coverage
- `test_atomic_claim_file.py` -- 25 unit tests covering all `AtomicClaimFile` methods and edge cases
- `test_startup_leader.py` -- 10 unit tests covering election, release, stale cleanup, and time-based expiry
- `test_main_unit_startup.py` -- 3 integration tests verifying the lifespan `_startup_sync` coroutine behavior
- `test_users_me_logout.py` -- 5 tests updated for the new `db.execute` pattern

#### Difficulties
- The atomic claim file needed `O_CREAT|O_EXCL` for race-free creation; standard `open()` has a TOCTOU window between checking file existence and creating it.
- Worker UUID must regenerate after `fork()` (uvicorn forks workers from a parent process). Lazy initialization with PID comparison handles this.

#### Risk Assessment
| Area | Risk | Notes |
|------|------|-------|
| Leader election | Low | Non-leaders simply skip startup sync; the existing 5-minute periodic sync job catches any gaps. If claim file fails, all workers fall through to original behavior. |
| Concurrent-write safety (users_me, middleware) | Low | Replaces ORM read-modify-write with single atomic SQL UPDATE. Strictly narrower failure surface. |
| IntegrityError handling | Low | Only catches the specific benign race on INSERT after UPDATE returns rowcount=0. All other exceptions propagate normally. |

### CODE REVIEW

File: studio/app/common/core/storage/atomic_claim_file.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+from datetime import datetime, timezone
+from typing import Optional, Tuple
+
+from studio.app.common.core.logger import AppLogger
+
+logger = AppLogger.get_logger()
+
+# Unique per-worker UUID, generated lazily on first access.
+# Lazy init ensures each forked worker gets its own UUID even when
+# the module is imported before fork (e.g. uvicorn --preload) (EC-25).
```
Comment (itutu-tienday) [inline studio/app/common/core/storage/atomic_claim_file.py:14]: What is (EC-25)?

File: studio/app/common/core/storage/atomic_claim_file.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    expiry = now + timedelta(minutes=expire_minutes)
+    return expiry.isoformat()
+
+
+class AtomicClaimFile:
+    """Reusable atomic file claim with stale detection.
+
+    Used by: startup leader election, download claim sentinels.
+    Pattern mirrors existing RemoteSyncLockFileUtil but uses O_CREAT|O_EXCL
+    instead of plain open() for race-free creation (EC-3).
```
Comment (itutu-tienday) [inline studio/app/common/core/storage/atomic_claim_file.py:42]: What is (EC-3)?

**Review by itutu-tienday — APPROVED** [review]
I approve it. Please test it in a real environment.

File: studio/app/common/core/storage/atomic_claim_file.py (hunk 1 above)
Comment (milesAraya) [reply studio/app/common/core/storage/atomic_claim_file.py:14]: Apologies this was Edge Case used during development, it will be removed
